File: src/main/java/ru/sfedu/teamselection/service/Oauth2UserService.java
package ru.sfedu.teamselection.service;

import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
import org.springframework.security.oauth2.core.oidc.user.DefaultOidcUser;
import org.springframework.stereotype.Service;
import ru.sfedu.teamselection.domain.User;
import ru.sfedu.teamselection.enums.Roles;
import ru.sfedu.teamselection.repository.UserRepository;

@Service
public class Oauth2UserService extends DefaultOAuth2UserService {

    private final UserRepository userRepository;

    public Oauth2UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public DefaultOidcUser loadUser(OAuth2UserRequest userRequest) {
        DefaultOidcUser user = (DefaultOidcUser) super.loadUser(userRequest);
        String email = user.getEmail();
        User userInDb = userRepository.findByEmail(email);

        if (userInDb == null) {
            User newUser = new User();
            newUser.setFio(user.getFullName());
            newUser.setEmail(email);
            newUser.setRole(Roles.USER);
            newUser.setIsEnabled(false);
            userRepository.save(newUser);
        }

        return user;
    }
}

File: src/main/kotlin/ru/sfedu/teamselection/service/Oauth2UserService.kt
package ru.sfedu.teamselection.service

import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest
import org.springframework.security.oauth2.core.oidc.user.DefaultOidcUser
import org.springframework.security.oauth2.core.user.OAuth2User
import org.springframework.stereotype.Service
import ru.sfedu.teamselection.domain.Users
import ru.sfedu.teamselection.enums.Roles
import ru.sfedu.teamselection.repository.UserRepository

@Service
class Oauth2UserService(
    private val userRepository: UserRepository
): DefaultOAuth2UserService() {
    override fun loadUser(userRequest: OAuth2UserRequest): OAuth2User {
        val user = super.loadUser(userRequest) as DefaultOidcUser
        val email = user.email
        val userInDb = userRepository.findByEmail(email)
        if (userInDb == null) {
            userRepository.save(Users(fio = user.fullName, email = email, role = Roles.USER, registered = false))
        }
        return user
    }
}